src/Fantomas/CodePrinter.fs

```diff
@@ -2591,7 +2591,11 @@ and genPat astContext pat =
     match pat with
     | PatOptionalVal(s) -> !- (sprintf "?%s" s)
     | PatAttrib(p, ats) -> genOnelinerAttributes astContext ats +> genPat astContext p
-    | PatOr(p1, p2) -> genPat astContext p1 +> sepNln -- "| " +> genPat astContext p2
+    | PatOr(p1, p2) ->
+        genPat astContext p1
+        +> enterNodeTokenByName pat.Range "BAR"
+        +> sepNone -- "| "
```

Comment:
As I suspected, changing the sepNln to sepNone leads to failing tests like:
```
  X match expressions [5ms]
  Error Message:
     Expected string length 117 but was 107. Strings differ at index 43.
  Expected: "...3 x =\n    match x with\n    | 1\n    | 2\n    | 3 -> printfn ..."
  But was:  "...3 x =\n    match x with\n    | 1| 2| 3 -> printfn "Found 1, ..."
  ----------------------------------------------^
```

Maybe try using creating a helper function in Context.fs.
The problem here is that the line comment will already introduced a new line I believe.
Try creating a function like:
```fsharp
let sepNlnUnlessLastEventIsNewline ctx =
     let alreadyHasNewline = lastWriteEventIsNewline ctx
    ( if alreadyHasNewline then sepNone else sepNln) ctx
```

Reply:
Awesome, thanks for that! The tests are all passing now. About the test, do you think it's sufficient?  Does it belong to CommentTest or PatternMatching ?

Reply:
I think both CommentTest and PatternMatching tests are good candidates.
In this case, I feel like CommentTest was the best choice. Although that is a bit subjective maybe.

Other test cases could be:
- the same test as you had but with [(* block comment *)](https://fsprojects.github.io/fantomas-tools/#/fantomas/preview?data=N4KABGBEDGD2AmBTSAuKAbRAXMAzMAHmALwA6ATqQHYQQC2AhltABaFgDuAllixdbTAAfMAEEwAeglguVdLMRg4dOoipZ%2BNWgAoAVGABG6WNADWS2CrU5dASk2CRAITABaAHxgAypcTb7lFoQIgD6bp4AcrBUiKSQcXEWVLhcAOYOtEKeAA7kiNlq8GAxHPIxGcGeAM4ssACu6EWIAI51DOhgCfGQkAA04FBV2FiyqVWoYADaAxCgghCQACRYAJ4FE5CyWH0ztEsAbu11yGjT87QADP3nCwCSVEjqXtkM0IgRdXQ7N2AALLsQAC6uwAvtdaHN5ktVus0Jt1N8oYtDuhjhMzjcAIzg86QAAKDFSiAA6lx4LxEedMQAmC4AsDAwRg3aQwTQtYnKAGWCwdCUvbIo6cjHnak4qFeRB0LhwYxUURYACiDwA8rgADIKfmCXDtIb0xm0ZmCVkCmGcyDc3naqCC1HC%2BkQADM4rZz1eiCciFwsDyBPIDFUWEQ5BtECw5GOBtB4tNC2WHI2Vr5rttKLRp0df1TC3dby9Pry6tgHBD0AYQ3u%2BxMTC40TDYAjUfOhogxoh9PZsK5PJTnbtGamWYArDmoHnPd7fYgAKrZArkcuVqjV8sjetj3XofUtmMs-vmpO9m0HIXorMANjHkAnBenAGF0BWqvfolUm9AsL6G1ud-NW2A7azAeiZwsmJ4Dg6PwAOzXreU55AAslKBghj%2BeqINGTKxiB3aWseqanva54-AAHHBLz5ghiCvnK6Hbphu7YfuuIJnh4GEZBJE3AAnBRHqiLgwbkK%2BKgMA2TaMf%2Be4mrhFocf26ZQViVxZjelGToWiCStKsoblmv5SYIAFAWAca2oeYEEYpZ6Zj8mLYmpE6CcJOkyry%2Bk-JJWFGjhrGWT21qcUp3FUmKan3I8WAqlQAAq5ArKSFKbhhPltn5SIBfhQU2cRdlYi6TkaaI5D1A8AAiiDyNKwkSZGRm0CZGVsmxFpbBBIX5VSvzXohDAELcuCxSwaiKgxXi1OQWCkuSLANr8dJMb5LGZaBUDtcFtlDvZo5qX1A3JFwBAqguTC%2BoqBC5IgVRVHWVANsOi3Scxsn%2BWt8LbJteXbViV57f1ABKiBwOQ8AzclWYLWlgHNWa70bblg4ivMmKwf9BAlQGKwquQmrvuDc1jlDS3pStLVZQjrGdT9VLkej6rYE4sjwKMBPzU9xkyR2b3sdZVNbcjgiYnxe0NCMZSesYZhOAGZjYFUMVeIGNG8p890pQx0OmeZXbyXzSLU4LtC0teESIKUChelgpZqLFHKVSkVA8HdogPMhdCoeQ4wa3%2BnMvdzq28zl-PfUbEDUo5Pw3hGMpYIhCDID7DVAqCAyAuCmxVAAYrdEyGSAIJAA) instead of a line comment.
- [Newlines between the or clauses](https://fsprojects.github.io/fantomas-tools/#/fantomas/preview?data=N4KABGBEDGD2AmBTSAuKAbRAXMAzMAHmALwA6ATqQHYQQC2AhltABaFgDuAllixdbTAAfMAEF%2BNWhMEiAQmAC0APjABlWHUQAKAJTSplSRBEBhRSoBysKogn7jYAPrmwVm-agAacFADO2LC4qAHNfVDAAbR8IUEEISAASLABPAAdkNEggrEhvOKgEgDcGdABXDMjouIAGPPyoAEkqJCosVVSGaEQLUrpcqsEAFgGwAF0qgF86mJHElPTwrNb%2B-MTisoqo%2BrAARmm4yAAFBmDEAHUueF4V%2Bp2AJmqR8cEpqtiDpLSKyAAjWFh0DdBGsSuVwlt6nd9sDVIg6Fw4OhrKIsABRZoAeVwABkgshobRcCV-E9Jvt3sDPgtMn8AUDaCCNuCRhAAMwE%2BLtTqIWSIXCwciIY7kBiaLCIcj0wRYcjlUkvcmzKnfWmAgmMsFoCH5QYcqBcrq8-mC7GwDgS6AMfxNQqwS2BaxS2gyuX5Z60V6CCkM5WLVVOjWbFlgACsesgBp5fIFiAAqql0uRLdaqLb7VxHXqiegSW6yW8lfMVf81Ur1prKtsAGzhyNGmMmdBW3wmay%2BF3QLACp0QbO5uLuiCe2je%2BK%2BmklgNFUFB7YAdlrHUN0cFAFk4T8JT28MTEPKPYrVuOoP71dOmVrgwAORfc%2BuC1tIqjbvt7vMKgtHot%2Bydn8uz%2BoAE5by6URcHFchWzoRhtxdN8B3zL1Cy%2BH86T-GdmW2HZamDCMlyjY1EFheFEUzYNX33IdDw%2Bb8JzQssMMvLC9lwyMwIg4iEQBMjtjgyiwGHGYvxQujSyPf9MNuKFcKaFosAxKgABVyGSC5rizXd%2BMEsBRwKWiT1-BiL0rW52VY-DRHIWBSmaAARRB0C4eEINg2V4MEQcBOoyl9KWHJ0OM7U4h2XVcNXBgCAaXBFJYRAqFRHMiJYAUsAuK4WG3QZHnfA9PxokSoGyKcJKY24wzCiKmlwLgCAxRMmAFVECFSQVfF8DNnz1ENsoQj8kOE6lCuWAKKyCwQdhrCqCAAJUQOByHgNL1ODLKtO8n1fKKkaAPyHYFymyyRWSDFyFxdsloyvVVpyqi8p8gq-OKxiTN2m8puxbBZCCeAgmCC7Mp6jzEJHZDBt%2BQzxOesbaB2YCwtKdBAkcmxZCRaAAGtZBFDHsF8BTVFFRBH16TryM0m6vLujaHtPIzRuDB5wwsRAOGRqMsHNOLFK%2BezqqoHgOtEZp1zoTdyDCDTErWqmx182nIcChmWO2CMZQRLBVwQfEyalimJh8UY6iyXwADF2vCCiJiAA).

I believe the fix you have will suffice for those cases as well, but extra tests make sure that we keep supporting the correct behaviour in future releases.